### DISCUSSION TITLE
How to modify file response (`onFileDownloadRequest`)

### DISCUSSION THREAD

**Discussion by crossbeamer** [pocketbase/pocketbase#5236] [Q&A] [answered]
`onFileDownloadRequest` is introduced as an event I could use to modify the "file response" before the client reads it. It's unclear to me how to go about doing this. I'm using the embedded JS engine and are using S3 for storage, though would eventually need to do this against files stored locally as well.

Do I need to read the file from S3 somehow? Or did PB do that before calling the event handler? Is the blob available through `theEvent.httpContext.Response` somewhere? Is using `$app.newFilesystem().getFile(...)` the right approach instead?

I need something here to point me in the right direction. Anyone familiar with this?

**Comment by ganigeorgiev** [comment]
_[Don't ask to ask, just ask](https://dontasktoask.com/)_

It would help to elaborate what actually you are trying to achieve (_how to write a custom response? how to modify the file before serving it? how to return a different file?_).

The `*Request` hooks expose `e.httpContext` which allows you to write a different response body (see [JSVM Routing](https://pocketbase.io/docs/js-routing/); `c` in the examples is the same as `e.httpContext`).

**Reply by crossbeamer** [reply]
> It would help to elaborate what actually you are trying to achieve (_how to write a custom response? how to modify the file before serving it? how to return a different file?_).

How to modify it. I want to write the file to disk, do something to it externally (via `$os`), read it back, and then send it to the client. And then in another scenario I'd like to do byte level manipulations on the JS side before sending it over.

From the event description, I was imagining I'd have an easy access to the file contents but failed to find any in the JSVM reference, or failed to properly apply what I thought could be it. Writing an arbitrary response body is very well explained and that part is clear to me.

**Reply by ganigeorgiev** [reply, marked as answer]
I want to help but I really don't understand what is the end goal here and I have no idea why users refuse to elaborate on their use case. You want to modify the file in place every time a download request is being made? _"I'd like to do byte level manipulations"_ - this doesn't really clarify anything, what type of manipulations?

Also note that the JSVM currently lacks helpers to operate with `io.Reader`/`io.Writer` interfaces so I'm not sure whether what is available at the moment will be sufficient for your use case and you may have to switch to Go.

I'm too tired for this type of interactions and will stop engage further but if you want to modify the file you need first to read it via the app filesystem abstraction. The event hook contains only information about the file that is going to be served (ex. `e.servedPath`). The actual file response write happens as last step, in the hook finalizer (this is to avoid unnecessary reads in case you want to prevent access to the file and return an error or read only portion of it in case the client sent a `Range` header). Here is a sample pseudo-code:

```js
onFileDownloadRequest((e) => {
    let fsys;
    try {
        fsys = $app.newFilesystem();

        let file;
        try {
            file = fsys.getFile(e.servedPath)

            // read it as string or use some of the blob.Reader methods
            // (https://pocketbase.io/jsvm/interfaces/blob.Reader.html)
            let content = readerToString(file)
            
            // change content....

            // "replace"
            fsys.upload(content, e.servedPath)
        } finally {
            file.close()
        }
    } finally {
        fsys.close();
    }


    // if you want to return a new file change e.servedPath to point to the new one or call
    // fsys.serve(e.httpContext.response(), e.httpContext.request(), filekey, name)
    // (https://pocketbase.io/jsvm/interfaces/filesystem.System.html#serve)
})
```

**Reply by crossbeamer** [reply]
`readerToString($app.newFilesystem().getFile(...))` is what I was after. Thank you.

I would have wanted to start the thread with a remark telling you to keep away from this but couldn't phrase it in a way that seemed appropriate, so I didn't. The *Q&A* section is presented as a place to ask **the community** for help and I wouldn't really expect the project maintainer to waste their time on every little question, especially if they felt the question wasn't specific enough or was otherwise unclear. Please don't burn out.
